File: OneDrive/Desktop/Dev-gpt-vms/backend/routes/maintenance.js
const express = require("express");
const router = express.Router();
const db = require("../models/MaintenanceTask");

// GET all tasks
router.get("/", (req, res) => {
  db.all("SELECT * FROM maintenance_tasks", [], (err, rows) => {
    if (err) return res.status(500).json({ error: err.message });
    res.json(rows);
  });
});

// GET task by id
router.get("/:id", (req, res) => {
  db.get("SELECT * FROM maintenance_tasks WHERE id = ?", [req.params.id], (err, row) => {
    if (err) return res.status(500).json({ error: err.message });
    if (!row) return res.status(404).json({ error: "Task not found" });
    res.json(row);
  });
});

// POST create task
router.post("/", (req, res) => {
  const { vesselId, task, type, scheduledDate, status, cost, notes } = req.body;
  db.run(
    "INSERT INTO maintenance_tasks (vesselId, task, type, scheduledDate, status, cost, notes) VALUES (?, ?, ?, ?, ?, ?, ?)",
    [vesselId, task, type, scheduledDate, status || "pending", cost, notes],
    function (err) {
      if (err) return res.status(500).json({ error: err.message });
      res.json({ id: this.lastID, vesselId, task, type, scheduledDate, status, cost, notes });
    }
  );
});

// PUT update task
router.put("/:id", (req, res) => {
  const { vesselId, task, type, scheduledDate, status, cost, notes } = req.body;
  db.run(
    "UPDATE maintenance_tasks SET vesselId=?, task=?, type=?, scheduledDate=?, status=?, cost=?, notes=? WHERE id=?",
    [vesselId, task, type, scheduledDate, status, cost, notes, req.params.id],
    function (err) {
      if (err) return res.status(500).json({ error: err.message });
      if (this.changes === 0) return res.status(404).json({ error: "Task not found" });
      res.json({ id: req.params.id, vesselId, task, type, scheduledDate, status, cost, notes });
    }
  );
});

// DELETE task
router.delete("/:id", (req, res) => {
  db.run("DELETE FROM maintenance_tasks WHERE id=?", [req.params.id], function (err) {
    if (err) return res.status(500).json({ error: err.message });
    if (this.changes === 0) return res.status(404).json({ error: "Task not found" });
    res.json({ message: "Task deleted" });
  });
});

module.exports = router;

File: OneDrive/Desktop/Dev-gpt-vms/backend/routes/maintenance.py
from fastapi import APIRouter, Depends, HTTPException, UploadFile, File
from sqlalchemy.orm import Session
from app.database import get_db
from backend.models.maintenance import MaintenanceTask
import shutil, os

router = APIRouter(prefix="/api/maintenance", tags=["maintenance"])

UPLOAD_DIR = "uploads/maintenance"
os.makedirs(UPLOAD_DIR, exist_ok=True)

@router.get("/")
def get_tasks(db: Session = Depends(get_db)):
    return db.query(MaintenanceTask).all()

@router.get("/{task_id}")
def get_task(task_id: int, db: Session = Depends(get_db)):
    task = db.query(MaintenanceTask).filter(MaintenanceTask.id == task_id).first()
    if not task:
        raise HTTPException(status_code=404, detail="Task not found")
    return task

@router.post("/")
def create_task(task: dict, db: Session = Depends(get_db)):
    new_task = MaintenanceTask(
        task_name=task.get("task_name"),
        description=task.get("description"),
        vessel_id=task.get("vessel_id"),
        assigned_to=task.get("assigned_to"),
        due_date=task.get("due_date"),
        status=task.get("status", "Pending"),
        cost=task.get("cost"),
        remarks=task.get("remarks"),
    )
    db.add(new_task)
    db.commit()
    db.refresh(new_task)
    return new_task

@router.put("/{task_id}")
def update_task(task_id: int, task_data: dict, db: Session = Depends(get_db)):
    task = db.query(MaintenanceTask).filter(MaintenanceTask.id == task_id).first()
    if not task:
        raise HTTPException(status_code=404, detail="Task not found")
    for key, value in task_data.items():
        setattr(task, key, value)
    db.commit()
    db.refresh(task)
    return task

@router.delete("/{task_id}")
def delete_task(task_id: int, db: Session = Depends(get_db)):
    task = db.query(MaintenanceTask).filter(MaintenanceTask.id == task_id).first()
    if not task:
        raise HTTPException(status_code=404, detail="Task not found")
    db.delete(task)
    db.commit()
    return {"detail": "Task deleted"}

@router.post("/{task_id}/evidence")
def upload_evidence(task_id: int, file: UploadFile = File(...), db: Session = Depends(get_db)):
    task = db.query(MaintenanceTask).filter(MaintenanceTask.id == task_id).first()
    if not task:
        raise HTTPException(status_code=404, detail="Task not found")

    file_path = os.path.join(UPLOAD_DIR, file.filename)
    with open(file_path, "wb") as buffer:
        shutil.copyfileobj(file.file, buffer)

    task.evidence_file = file_path
    db.commit()
    db.refresh(task)
    return {"filename": file.filename, "path": file_path}
